File: Pythoncode/1143.最长公共子序列.py
class Solution:
    def longestCommonSubsequence(self, text1: str, text2: str) -> int:
        d = {} # d[(i, j)] = text1[:i] 和 text2[:j]之间的最长公共子序列
        maxcomsublen = 0
        for i in range(len(text1)+1):
            for j in range(len(text2)+1):
                maxcomsublen = max(maxcomsublen, self.getMaxComSublen(text1, text2, i, j, d))
        return maxcomsublen

    def getMaxComSublen(self, text1, text2, i, j, d):
        if i == 0 or j == 0:
            return 0
        if (i, j) in d:
            return d[(i, j)]
        if text1[:i][-1] != text2[:j][-1]:
            d[(i, j)] = max(self.getMaxComSublen(text1, text2, i-1, j, d), self.getMaxComSublen(text1, text2, i, j-1, d))
        else:
            d[(i, j)] = 1 + self.getMaxComSublen(text1, text2, i-1, j-1, d)
        return d[(i, j)]

File: Cppcode/1143.最长公共子序列.cpp
#include <string>
#include <vector>

using namespace std;

class Solution {
public:
    int longestCommonSubsequence(string text1, string text2) {
        vector<vector<int>> dp(text1.size(), vector<int>(text2.size(), 0));
        for (auto i = 0; i < text1.size(); i++)
        {
            for (auto j = 0; j < text2.size(); j++)
            {
                auto a = i-1>=0 ? dp[i-1][j] : 0;
                auto b = j-1>=0 ? dp[i][j-1] : 0;
                auto c = text1[i] == text2[j] ? 1 : 0;
                if (i-1>=0 && j-1>=0) c += dp[i-1][j-1];
                dp[i][j] = max(max(a, b) ,c);
            }
        }
        return dp.back().back();
    }
};